README
NAME
    ExtUtils::ParseXS - converts Perl XS code into C code

SYNOPSIS
      use ExtUtils::ParseXS qw(process_file);
  
      process_file( filename => 'foo.xs' );

      process_file( filename => 'foo.xs',
                    output => 'bar.c',
                    'C++' => 1,
                    typemap => 'path/to/typemap',
                    hiertype => 1,
                    except => 1,
                    prototypes => 1,
                    versioncheck => 1,
                    linenumbers => 1,
                    optimize => 1,
                    prototypes => 1,
                  );
    =head1 DESCRIPTION

    "ExtUtils::ParseXS" will compile XS code into C code by embedding the
    constructs necessary to let C functions manipulate Perl values and
    creates the glue necessary to let Perl access those functions. The
    compiler uses typemaps to determine how to map C function parameters and
    variables to Perl values.

    The compiler will search for typemap files called *typemap*. It will use
    the following search path to find default typemaps, with the rightmost
    typemap taking precedence.

            ../../../typemap:../../typemap:../typemap:typemap

EXPORT
    None by default. "process_file()" may be exported upon request.

FUNCTIONS
    process_xs()
        This function processes an XS file and sends output to a C file.
        Named parameters control how the processing is done. The following
        parameters are accepted:

        C++ Adds "extern "C"" to the C code. Default is false.

        hiertype
            Retains "::" in type names so that C++ hierachical types can be
            mapped. Default is false.

        except
            Adds exception handling stubs to the C code. Default is false.

        typemap
            Indicates that a user-supplied typemap should take precedence
            over the default typemaps. A single typemap may be specified as
            a string, or multiple typemaps can be specified in an array
            reference, with the last typemap having the highest precedence.

        prototypes
            Generates prototype code for all xsubs. Default is false.

        versioncheck
            Makes sure at run time that the object file (derived from the
            ".xs" file) and the ".pm" files have the same version number.
            Default is true.

        linenumbers
            Adds "#line" directives to the C output so error messages will
            look like they came from the original XS file. Default is true.

        optimize
            Enables certain optimizations. The only optimization that is
            currently affected is the use of *target*s by the output C code
            (see perlguts). Not optimizing may significantly slow down the
            generated code, but this is the way xsubpp of 5.005 and earlier
            operated. Default is to optimize.

        inout
            Enable recognition of "IN", "OUT_LIST" and "INOUT_LIST"
            declarations. Default is true.

        argtypes
            Enable recognition of ANSI-like descriptions of function
            signature. Default is true.

        s   I have no clue what this does. Strips function prefixes?

    errors()
        This function returns the number of [a certain kind of] errors
        encountered during processing of the XS file.

AUTHOR
    Based on xsubpp code, written by Larry Wall.

    Maintained by:

    *   Ken Williams, <ken@mathforum.org>

    *   David Golden, <dagolden@cpan.org>

COPYRIGHT
    Copyright 2002-2009 by Ken Williams, David Golden and other
    contributors. All rights reserved.

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    Based on the ExtUtils::xsubpp code by Larry Wall and the Perl 5 Porters,
    which was released under the same license terms.

SEE ALSO
    perl, ExtUtils::xsubpp, ExtUtils::MakeMaker, perlxs, perlxstut.



Changes
Revision history for Perl extension ExtUtils::ParseXS.

2.2207 - 

2.2206 - Sun Jul  4 15:43:21 EDT 2010

 Bug fixes:

 - Make xsubpp accept the _ prototype (RT#57157) [Rafael Garcia-Suarez]

 - INCLUDE_COMMAND portability fixes for VMS (RT#58181) [Craig Berry]

 - INCLUDE_COMMAND fixes to detect non-zero exit codes (RT#52873)
   [Steffen Mueller]

2.2205 - Wed Mar 10 18:15:36 EST 2010

 Other:

 - No longer ships with Build.PL to avoid creating a circular dependency

2.2204 - Wed Mar 10 14:23:52 EST 2010

 Other:

 - Downgraded warnings on using INCLUDE with a command from "deprecated"
   to "discouraged" and limited it to the case where the command includes
   "perl" [Steffen Mueller]

2.2203 - Thu Feb 11 14:00:51 EST 2010

 Bug fixes:

 - Build.PL was not including ExtUtils/xsubpp for installation.  Fixed
   by subclassing M::B::find_pm_files to include it [David Golden]

2.2202 - Wed Jan 27 15:04:59 EST 2010

 Bug fixes:

 - The fix to IN/OUT/OUTLIST was itself broken and is now fixed.
   [Reported by Serdar Dalgic; fix suggested by Rafael Garcia-Suarez]

   We apologize for the fault in the regex. Those responsible 
   have been sacked.

2.2201 Mon Jan 25 16:12:05 EST 2010

 Bug fixes:

 - IN/OUT/OUTLIST, etc. were broken due to a bad regexp.  [Simon Cozens]

2.22 - Mon Jan 11 15:00:07 EST 2010

 No changes from 2.21_02

2.21_02 - Sat Dec 19 10:55:41 EST 2009

 Bug fixes:

 - fixed bugs and added tests for INCLUDE_COMMAND [Steffen Mueller]

2.21_01 - Sat Dec 19 07:22:44 EST 2009

 Enhancements:

 - New 'INCLUDE_COMMAND' directive [Steffen Mueller]

 Bug fixes:

 - Workaround for empty newXS macro found in P5NCI [Goro Fuji]

2.21 - Mon Oct  5 11:17:53 EDT 2009

 Bug fixes:
 - Adds full path in INCLUDE #line directives (RT#50198) [patch by "spb"]

 Other:
 - Updated copyright and maintainer list

2.20_07 - Sat Oct  3 11:26:55 EDT 2009

 Bug fixes:
 - Use "char* file" for perl < 5.9, not "char[] file"; fixes mod_perl
   breakage due to prior attempts to fix RT#48104 [David Golden]

2.20_06 - Fri Oct  2 23:45:45 EDT 2009

 Bug fixes:
 - Added t/typemap to fix broken test on perl 5.6.2 [David Golden]
 - More prototype fixes for older perls [Goro Fuji]
 - Avoid "const char *" in test files as it breaks on 5.6.2 [Goro Fuji]

 Other:
 - Merged changes from 2.2004 maintenance branch (see 2.200401 to 2.200403)
   [David Golden]

2.20_05 - Sat Aug 22 21:46:56 EDT 2009

 Bug fixes:
 - Fix prototype related bugs [Goro Fuji]
 - Fix the SCOPE keyword [Goro Fuji]

2.200403 - Fri Oct  2 02:01:58 EDT 2009

 Other:
 - Removed PERL_CORE specific @INC manipulation (no longer needed)
   [Nicholas Clark]
 - Changed hard-coded $^H manipulation in favor of "use re 'eval'"
   [Nicholas Clark]

2.200402 - Fri Oct  2 01:26:40 EDT 2009

 Bug fixes:
 - UNITCHECK subroutines were not being called (detected in ext/XS-APItest
   in Perl blead) [reported by Jesse Vincent, patched by David Golden]

2.200401 - Mon Sep 14 22:26:03 EDT 2009

 - No changes from 2.20_04.


2.20_04 - Mon Aug 10 11:18:47 EDT 2009

 Bug fixes:
 - Changed install_dirs to 'core' for 5.8.9 as well (RT#48474)
 - Removed t/bugs.t until there is better C++ support in ExtUtils::CBuilder

 Other:
 - Updated repository URL in META file

2.20_03 - Thu Jul 23 23:14:50 EDT 2009

 Bug fixes:
 - Fixed "const char *" errors for 5.8.8 (and older) (RT#48104)
   [Vincent Pit]
 - Added newline before a preprocessor directive (RT#30673)
   [patch by hjp]

2.2002 - Sat Jul 18 17:22:27 EDT 2009

 Bug fixes:
 - Fix Makefile.PL installdirs for older perls

2.20_01 - Wed Jul  8 12:12:47 EDT 2009

 - Fix XSUsage prototypes for testing [Jan Dubois]

2.20 - Wed Jul  1 13:42:11 EDT 2009

 - No changes from 2.19_04

2.19_04 - Mon Jun 29 11:49:12 EDT 2009

 - Changed tests to use Test::More and added it to prereqs

 - Some tests skip if no compiler or if no dynamic loading

 - INTERFACE keyword tests skipped for perl < 5.8

2.19_03 - Sat Jun 27 22:51:18 EDT 2009

 - Released to see updated results from smoke testers

 - Fix minor doc typo pulled from blead

2.19_02 - Wed Aug  6 22:18:33 2008

 - Fix the usage reports to consistently report package name as well
   as sub name across ALIAS, INTERFACE and regular XSUBS. [Robert May]

 - Cleaned up a warning with -Wwrite-strings that gets passed into
   every parsed XS file. [Steve Peters]

 - Allow (pedantically correct) C pre-processor comments in the code
   snippets of typemap files. [Nicholas Clark]

2.19 - Sun Feb 17 14:27:40 2008

 - Fixed the treatment of the OVERLOAD: keyword, which was causing a C
   compile error. [Toshiyuki Yamato]

2.18 - Mon Jan 29 20:56:36 2007

 - Added some UNITCHECK stuff, which (I think) makes XS code able to
   do UNITCHECK blocks. [Nicholas Clark]

 - Changed 'use re "eval";' to 'BEGIN { $^H |= 0x00200000 };' so we
   can compile re.xs in bleadperl. [Yves Orton]

 - Fix an undefined-variable warning related to 'inout' parameter
   processing.

2.17 - Mon Nov 20 17:07:27 2006

 - Stacked $filepathname to make #line directives in #INCLUDEs work.
   [Nicholas Clark]

 - Sprinked dVAR in with dXSARGS, for God-(Jarkko)-knows-what
   reason. [Jarkko Hietaniemi]

 - Use printf-style formats in Perl_croak() for some significant
   savings in number of distinct constant strings in the linked
   binaries we create. [Alexey Tourbin]

 - Don't use 'class' as a variable name in the t/XSTest.xs module,
   since that's a keyword in C++. [Jarkko Hietaniemi]

2.16  Fri Sep 15 22:33:24 CDT 2006

 - Fix a problem with PREFIX not working inside INTERFACE
   sections. [Salvador Fandin~o]

2.15  Mon Oct 10 11:02:13 EDT 2005

 - I accidentally left out a README from the distribution.  Now it's
   auto-created from the main documentation in ExtUtils/ParseXS.pm.

2.14  Sat Oct  8 21:49:15 EDT 2005

 - The filehandle for the .xs file was never being properly closed,
   and now it is.  This was causing some Win32 problems with
   Module::Build's tests, which create a .xs file, process it with

Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.3603
use ExtUtils::MakeMaker;
WriteMakefile
(
          'NAME' => 'ExtUtils::ParseXS',
          'VERSION_FROM' => 'lib/ExtUtils/ParseXS.pm',
          'PREREQ_PM' => {
                           'Carp' => 0,
                           'Cwd' => 0,
                           'DynaLoader' => 0,
                           'Exporter' => 0,
                           'ExtUtils::CBuilder' => 0,
                           'File::Basename' => 0,
                           'File::Spec' => 0,
                           'Symbol' => 0,
                           'Test::More' => '0.47'
                         },
          'INSTALLDIRS' => ($] < 5.008009 ? 'site' : 'perl'),
          'EXE_FILES' => [],
          'PL_FILES' => {}
        )
;


devtools/bleadcheck.pl
#!/usr/bin/perl

# A script to check a local copy against bleadperl, generating a blead
# patch if they're out of sync.  The path to bleadperl is required.
# An optional directory argument will be chdir()-ed into before comparing.

use strict;
my $blead = shift @ARGV
  or die "Usage: $0 <bleadperl-src> [ExtUtils-ParseXS-src]\n";

chdir shift() if @ARGV;


diff( "$blead/lib/ExtUtils/ParseXS.pm", "lib/ExtUtils/ParseXS.pm");
diff( "$blead/lib/ExtUtils/xsubpp", "lib/ExtUtils/xsubpp");
diff( "$blead/lib/ExtUtils/ParseXS", "lib/ExtUtils/ParseXS",
      qw(t Changes .svn) );

diff( "$blead/lib/ExtUtils/ParseXS/t", "t",
      qw(.svn) );

######################
sub diff {
  my ($first, $second, @skip) = @_;
  local $_ = `diff -ur $first $second`;

  for my $x (@skip) {
    s/^Only in .* $x\n//mg;
  }
  print;
}


devtools/bump_version.pl
#!/usr/bin/env perl

# NOTE: we run this immediately *after* a release so that any reports
# against svn are obvious

use strict;
use warnings; # hi, Jim!

use lib 'lib';
use Module::Build;

use Tie::File;

eval { require File::Find::Rule } or
  die "$0 requires File::Find::Rule. Please install and try again.\n";

my $current = Module::Build->new_from_context(quiet => 1)->dist_version;

# Get version from command line or prompt
my $version = shift;
unless($version) {
  my $default = $current;

  # try to construct a reasonable default automatically
  $default =~ s/(\d+)$// or
    die "Usage: $0 VERSION\ncurrently: $current\n";
  my $end = $1;
  $default .= sprintf('%0'.length($end).'d', $end+1);

  local $| = 1;
  print "enter new version [$default]: ";
  chomp(my $ans = <STDIN>);
  $version = $ans ? $ans : $default;
  # TODO check for garbage in?
}

die "must bump forward! ($version < $current)\n"
  unless(eval $version >= eval $current);

# NEVER BUMP THESE $VERSION numbers
my @excluded = qw(
  lib/Module/Build/Version.pm
  lib/Module/Build/YAML.pm
);

# Get list of .pm files
my @pmfiles = File::Find::Rule->new->or(
  File::Find::Rule->name('*.pm'),
  File::Find::Rule->directory->name( qr/\.svn/ )->prune->discard
)->in( 'lib' );
my @scripts = File::Find::Rule->new()->or(
  File::Find::Rule->name('*'),
  File::Find::Rule->directory->name( qr/\.svn/ )->prune->discard
)->in( './scripts' );

# first start the new Changes entry
sub {
  my $file = 'Changes';
  open(my $fh, '<', $file) or die "cannot read '$file' $!";
  my @lines = <$fh>;
  my @head;
  while(@lines) {
    my $line = shift(@lines);
    if($line =~ m/^$current/ ) {
      # unreleased case -- re-bumping
      if($line =~ m/^$current(?: *- *)?$/) {
        print "Error parsing $file - found unreleased '$current'\n"; 
        local $| = 1;
        print "Are you sure you want to change the version number (y/n)? [n]:";
        chomp(my $ans = <STDIN>);
        if ( $ans !~ /^y/i ) {
          print "Aborting!\n";
          exit 1;
        }
        warn "Updating '$file'\n";
        open(my $ofh, '>', $file) or die "cannot write '$file' $!";
        print $ofh @head, "$version - \n", @lines;
        close($ofh) or die "cannot write '$file' $!";
        return;
      }
      if($line =~ m/^$current - \w/) {
        warn "Updating '$file'\n";
        open(my $ofh, '>', $file) or die "cannot write '$file' $!";
        print $ofh @head, "$version - \n", "\n", $line, @lines;
        close($ofh) or die "cannot write '$file' $!";
        return;
      }
      elsif($line =~ m/^$version(?: *- *)?$/) {
        # TODO should just be checking for a general number+eol case?
        die "$file probably needs to be reverted!";
      }
    }
    else {
      push(@head, $line);
    }
  }
  die "cannot find changes entry for current version ($current)!";
}->();

for my $file ( @pmfiles, @scripts ) {
  next if grep { $file eq $_ } @excluded;
  bump_version( $file, $version );
}


exit;

sub bump_version {
  my ( $file, $version ) = @_;
  my $o = tie my @lines, 'Tie::File', $file 
    or die "Couldn't tie '$file' for editing\n";
  $o->flock;

  # find line to change just like EU::MM::parse_version
  my $inpod = 0;
  for ( @lines ) {
      $inpod = /^=(?!cut)/ ? 1 : /^=cut/ ? 0 : $inpod;
      next if $inpod || /^\s*#/;
      next unless /(?<!\\)([\$*])(([\w\:\']*)\bVERSION)\b.*\=/;
      # TODO check that what we found matches $current?
      $_ = "\$VERSION = '$version';"; 
      warn "Updated $file\n";
      last;
  }

  undef $o; untie @lines;
  return;
}

# vi:ts=2:sw=2:et:sta


Build.PL
use Module::Build;
use Config;

my $class = Module::Build->subclass(code => <<'EOF');
  # add lib/ExtUtils/xsubpp
  sub find_pm_files {
    my $self = shift;
    my $pm_files = $self->SUPER::find_pm_files;
    my $xsubpp = $self->localize_file_path('lib/ExtUtils/xsubpp');
    $pm_files->{$xsubpp} = $xsubpp;
    return $pm_files
  }

  sub do_create_makefile_pl {
    my $self = shift;
    $self->SUPER::do_create_makefile_pl(fh => $fh);
    $self->do_system(qw(perl -pi -e), q{s/'INSTALLDIRS' => '\w+'/'INSTALLDIRS' => (\$] < 5.008009 ? 'site' : 'perl')/}, 'Makefile.PL');
  }
EOF

my $build = $class->new
(
  module_name => 'ExtUtils::ParseXS',
  license => 'perl',
  installdirs => ($] < 5.008009 ? 'site' : 'core'),
  auto_configure_requires => 0,
  requires => {
    'Cwd' => 0,
    'Exporter' => 0,
    'File::Basename' => 0,
    'File::Spec' => 0,
    'Symbol' => 0,
  },
  build_requires => {
    'Carp' => 0,
    'DynaLoader' => 0,
    'ExtUtils::CBuilder' => 0,
    'Test::More' => 0.47,
  },
  add_to_cleanup => ["t/XSTest.c", "t/XSTest$Config{obj_ext}", "t/XSTest.$Config{dlext}"],
  create_makefile_pl => 'traditional',
  create_readme => 1,
  meta_merge     => {
    resources => {
      bugtracker =>
        'http://rt.cpan.org/Dist/Display.html?Queue=ExtUtils-ParseXS',
      MailingList => 'mailto:perl5-porters@perl.org',
      repository  => 'git://github.com/dagolden/extutils-parsexs.git'
    }
  },
);

$build->create_build_script;

